# Drink Caffeine Content Analysis Project

## Project Introdocution

In this analysis we will delve into a topic that is very near and dear to my heart - caffeine! I have a caffeine addiction. More often than not I opt for energy drinks ala Red Bull, Monster, Rockstar, etc. However, when the weather gets cold I start to really enjoy coffee. I often worry that the amount of caffeine I ingest will one day come back to bite me in terms of health. Despite how much caffeine I consume on a weekly basis, I know shockingly little about the amount of caffeine in the drinks I enjoy so much. 

This data set contains information related to roughly 600 different drinks with caffeine and provides information for each related to caffeine content, volume, and calories. We will conduct some exploratory data analysis (EDA) and then attempt to find a subset of drinks which may be "less unhealthy" and could serve as some alternatives to sugar filled energy drinks and coffee.


## Import Libraries and Load the Dataset

In [1]:
# Import libraries used in the project
import numpy as np
import pandas as pd
import plotly.express as px
import plotly.io as pio



In [2]:
# import the caffeine dataset
caffeine = pd.read_csv(r'C:\Users\benwh\Documents\Projects\Caffeine\caffeine.csv')

## Data Familiarization

In [3]:
caffeine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 610 entries, 0 to 609
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   drink          610 non-null    object 
 1   Volume (ml)    610 non-null    float64
 2   Calories       610 non-null    int64  
 3   Caffeine (mg)  610 non-null    int64  
 4   type           610 non-null    object 
dtypes: float64(1), int64(2), object(2)
memory usage: 24.0+ KB


In [4]:
caffeine.head()

drink  Volume (ml)  Calories  Caffeine (mg)    type
0                 Costa Coffee   256.993715         0            277  Coffee
1  Coffee Friend Brewed Coffee   250.191810         0            145  Coffee
2           Hell Energy Coffee   250.191810       150            100  Coffee
3           Killer Coffee (AU)   250.191810         0            430  Coffee
4                 Nescafe Gold   250.191810         0             66  Coffee

We can see after running .info() and .head() that our data set is pretty straight forward - 610 observations with five fields covering drink name, volume, calories, caffeine content and drink type.

In [5]:
# Check for duplicate drink names
duplicatedrinks = caffeine[caffeine.duplicated(['drink'])]
duplicatedrinks

Empty DataFrame
Columns: [drink, Volume (ml), Calories, Caffeine (mg), type]
Index: []

We have no duplicate drink names in our dataset.

In [6]:
# Distribution of drink types in the dataset.
caffeine['type'].value_counts()

Energy Drinks    219
Coffee           173
Soft Drinks       90
Tea               66
Energy Shots      36
Water             26
Name: type, dtype: int64

Value_counts() reveals there are six unique drink types.

In [7]:
# Distribution of types of volume in the dataset.
caffeine['Volume (ml)'].value_counts()

354.882000    159
473.176000    125
236.588000     96
250.191810     35
59.147000      19
             ... 
428.815750      1
380.019475      1
370.260220      1
339.208045      1
329.744525      1
Name: Volume (ml), Length: 70, dtype: int64

The volume of drinks varies quite a bit, but more than half of all volumes are higher than 200 mL.

In [8]:
# Distribution of calorie contnet in the dataset.
caffeine['Calories'].value_counts()

0      200
5       32
10      30
160     22
140     21
      ... 
155      1
117      1
78       1
118      1
299      1
Name: Calories, Length: 98, dtype: int64

Calories per drink appear to varry quite a bit. Notably, just under half appear to be very low in calories.

In [9]:
# Distribution of caffeine content in the dataset.
caffeine['Caffeine (mg)'].value_counts()

80      37
100     34
160     33
0       28
120     23
        ..
1150     1
1325     1
728      1
1101     1
99       1
Name: Caffeine (mg), Length: 162, dtype: int64

While there are six types of drinks included in the dataset, there is a lot more variation among volume, calories, and caffeine content.

## Univariate Analysis 

### Summary Statistics for Volume (ml), Calories, and Caffeine (mg)

In [10]:
caffeine.describe()

Volume (ml)    Calories  Caffeine (mg)
count   610.000000  610.000000     610.000000
mean    346.543630   75.527869     134.693443
std     143.747738   94.799919     155.362861
min       7.393375    0.000000       0.000000
25%     236.588000    0.000000      50.000000
50%     354.882000   25.000000     100.000000
75%     473.176000  140.000000     160.000000
max    1419.528000  830.000000    1555.000000

In [11]:
caffeine.groupby(['type']).describe()

Volume (ml)                                                  \
                    count        mean         std         min         25%   
type                                                                        
Coffee              173.0  335.870855  159.716392   12.716605  236.588000   
Energy Drinks       219.0  388.971198  106.409997  236.588000  250.191810   
Energy Shots         36.0   57.742259   22.094888    7.393375   57.076855   
Soft Drinks          90.0  355.243454   41.509635  236.588000  354.882000   
Tea                  66.0  360.474080  167.002318  177.441000  236.588000   
Water                26.0  394.590111   99.702740  236.588000  343.791938   

                                             Calories              ...  \
                    50%      75%         max    count        mean  ...   
type                                                               ...   
Coffee         325.3085  473.176  1419.52800    173.0   73.497110  ...   
Energy Drinks  354.8820  473.176   751.16690    219.0   86.671233  ...   
Energy Shots    59.1470   59.147   125.98311     36.0   16.500000  ...   
Soft Drinks    354.8820  354.882   591.47000     90.0  111.111111  ...   
Tea            236.5880  473.176   946.35200     66.0   52.757576  ...   
Water          354.8820  473.176   591.47000     26.0   11.538462  ...   

                             Caffeine (mg)                                \
                  75%    max         count        mean         std   min   
type                                                                       
Coffee         120.00  830.0         173.0  200.589595  248.222165   2.0   
Energy Drinks  149.50  320.0         219.0  147.867580   76.734535   0.0   
Energy Shots    25.00  100.0          36.0  193.416667   79.535931  75.0   
Soft Drinks    160.00  320.0          90.0   33.677778   24.915961   0.0   
Tea            107.50  299.0          66.0   55.863636   39.333637   0.0   
Water            3.75  110.0          26.0   53.730769   34.060602   0.0   

                                              
                  25%    50%     75%     max  
type                                          
Coffee         100.00  145.0  200.00  1555.0  
Energy Drinks   80.00  135.0  182.50   400.0  
Energy Shots   120.00  192.5  241.25   350.0  
Soft Drinks      9.25   37.0   47.75   102.0  
Tea             29.25   45.0   68.00   165.0  
Water           35.00   60.0   73.75   120.0  

[6 rows x 24 columns]

### Distribution of Drink Types

In [12]:
caffeine_gb = caffeine.groupby(["type"]).count()
fig = px.bar(caffeine_gb, 
             x=caffeine_gb.index, 
             y="drink", 
             title='Count of Drink Types')
fig.update_layout(xaxis={'categoryorder':'total descending'})
fig.show()

### Histogram - Volume (ml)

In [13]:
px.histogram(caffeine, 
             x = 'Volume (ml)', 
             title='Distribution of Volume')

The right tail of the histogram for Volume (ml) is pretty long. Let's zoom in and look at a histogram of volumes > 800 ml.

In [14]:
large_volume = caffeine[(caffeine['Volume (ml)'] > 800)]
large_volume

drink  Volume (ml)  Calories  Caffeine (mg)  \
29   Starbucks Bottled Iced Coffee     1419.528       240            640   
530            McDonalds Sweet Tea      946.352       160            100   

       type  
29   Coffee  
530     Tea

Our large right-sided tail is only made of two drinks.

### Histogram - Calories

In [15]:
px.histogram(caffeine, 
             x = 'Calories', 
             title='Distribution of Calories')

Similar to to Volume (ml), we have a large right-sided tail for Calories. Let's take a closer look and filter for drinks with >= 400 calories.

In [16]:
high_calories = caffeine[(caffeine['Calories'] >= 400)]
high_calories

drink  Volume (ml)  Calories  Caffeine (mg)  \
13                 Dare Iced Coffee   500.087885       429            160   
30  Baskin Robbins Cappuccino Blast   709.764000       470            234   
37             Arby's Jamocha Shake   473.176000       830             12   
81   Big Train Java Chip Ice Coffee   354.882000       410             49   

      type  
13  Coffee  
30  Coffee  
37  Coffee  
81  Coffee

The long right-sided tail for calories is made up of four drinks that all look like they are a liquid version of ice cream.

### Histogram - Caffeine (mg)

In [17]:
px.histogram(caffeine, 
             x = 'Caffeine (mg)', 
             title='Distribution of Caffeine', 
             template='plotly_white')

Again we have a long, right-sided tail. This time for Caffeine. Let's zoom in on the tail by looking at any drink with > 400 mg of caffeine.

In [18]:
high_caffeine = caffeine[(caffeine['Caffeine (mg)'] > 400)]

px.histogram(high_caffeine, 
             x = 'Caffeine (mg)', 
             nbins=10, 
             title='Distribution of Drinks With Caffeine Content Higher Than 400 mg', 
             template='plotly_white')

### Caffeine by Drink Type

### Univariate Analysis - Takeaways

- A bit over half of the dataset is made up of Enery Drink and Coffee drink types
- Most volumes fall between 200 and 475 ml
- More than half of the drinks in the dataset appear to be low calorie (< 100 calories
- Most drinks have less than 200 mg caffeine
- ~ 75% drinks have a caffeine concentration of .75 or less

## Bivariate Analysis

### Box Plot for Caffeine Across all Drink Types

In [19]:
fig = px.box(caffeine, 
            x='type', 
            y='Caffeine (mg)',
            title='Caffeine Content By Drink Type',
            labels={'type': 'Type'},
            template='plotly_white',
            hover_data=['drink', 'Volume (ml)']
            )
fig

Coffee has numerous outlying drinks where caffeine content is very high. Additionally, Coffee drinks appear to be a bit higher in terms of caffeine content than Energy Drinks. However, Energy Shots are higher on average than both Coffee and Energy Drinks. 

### Volume x Caffeine

In [20]:
px.scatter(caffeine,
            x='Volume (ml)', 
            y='Caffeine (mg)',
            color='type',  
            title='Volume x Caffeine',
            height = 800,
            template='plotly_white',
            hover_data=['drink']
            )

The coffee outliers can be seen here as well and it looks like they are mostly among drinks of the same volume. Energy shots are clustered around 100 ml or lower but have similar caffeine content to many enery drinks and coffee. Tea, Water, and Soft Drinks look to lower caffeine content. 

## Box Plot for Caffeine Concentration Across all Drink Types

We're going to take another look at caffeine distribution across all drink types, but this time we'll use caffeine concentration which is equal to Caffeine (mg) / Volume (ml). This will give us a better idea of which drink types pack a caffeine punch relative to the volume of the drink.

In [ ]:
# create Caffeine Concentration column which = Caffeine (mg) / Volume (ml)
caffeine['Caffeine Concentration'] = caffeine['Caffeine (mg)'] / caffeine['Volume (ml)']

px.box(caffeine,
        x='type',
        y='Caffeine Concentration',
        title='Caffeine Concentration By Drink Type',
        labels={'type': 'Type'},
        height=800,
        template='plotly_white',
        hover_data=['drink']
        )

The caffeine concentration of energy shots is clearly much higher than any other drink type. Between Coffee and Energy drinks, Coffee appears to have a higher caffeine concentration. Additionally, just like with the previous box plot showing caffeine content for all drink types, there are numerous outliers for coffee.


## Healthier Options

Now that we have a better understanding of the data set and have created a caffeine concentration value for each drink, lets try to see if we can isolate some potentially healthier options. In this data set, we only have calories as a data point which could be extrapolated to interpret whether something is healthy or not, so we will run with that. 

Let's take filter for a subset of drinks that have 100 or less calories

In [ ]:
healthy_caffeine = caffeine[(caffeine['Calories'] <= 100) & (caffeine['type'] != 'Energy Shots')].sort_values(by=['Caffeine Concentration'], ascending=False).head(20)
healthy_caffeine

## Which Tea and Water Drinks Have the Highest Caffeine Content

As much as most people love caffeine, it's hard not to feel a little bit guilty indulging in a caffeinated drink because unless you're drinking black coffee, drinks with caffeine tend to come with a lot of other chemicals included in them.

Let's take a look at the Caffeine dataset but only at Tea and Water drinks to see if we can find some drinks with a lot of caffeine

In [23]:
# filter for only Tea and Water
tw_caffeine = caffeine[(caffeine['type'] == 'Tea') | (caffeine['type'] == 'Water')].sort_values(by=['Caffeine Concentration'], ascending=False).head(10)

tw_caffeine

drink  Volume (ml)  Calories  Caffeine (mg)   type  \
581  Zest Highly Caffeinated Tea    236.58800         0            150    Tea   
556          Fast Lane Black Tea    236.58800         0            110    Tea   
558                    HICAF Tea    236.58800         0            110    Tea   
584             Perrier Energize    250.19181        35             99  Water   
528      Inko's White Tea Energy    458.38925       100            165    Tea   
580          YMateina Yerba Mate    236.58800         0             80    Tea   
598         Arti Sparkling Water    354.88200         0            120  Water   
527    Guayaki Canned Yerba Mate    458.38925       120            150    Tea   
542           Taiwanese Milk Tea    473.17600       299            151    Tea   
561    Lipton Natural Energy Tea    236.58800         0             75    Tea   

     Caffeine Concentration  
581                0.634014  
556                0.464943  
558                0.464943  
584                0.395696  
528                0.359956  
580                0.338141  
598                0.338141  
527                0.327233  
542                0.319120  
561                0.317007